instadam/project.py

```diff
@@ -319,3 +320,74 @@ def update_user_permission(project_id):
         db.session.commit()
 
     return construct_msg('Permission added successfully'), 200
+
+
+@bp.route('/project/<project_id>/request', methods=['POST'])
+@jwt_required
+def request_permission(project_id):
+    """ Request for specific permission to a project
+
+        Request for specified privilege (READ_WRITE or READ_ONLY) to project
+        upon receiving a `POST` request to the `/project/<project_id>/request`
+        entry point. User must be signed in. User must provide a `message_type`
+        to specify the privilege he or she is requesting for.
+
+        `message_type` takes a string of two values: `r_request` or `rw_request`,
+        where `r_request` is `READ_ONLY` and `rw_request` is `READ_WRITE`
+
+        The request message will be sent to all admins of the project. Note that
+        an annotator can also send a request of 'rw' permission. The decision of
+        whether upgrading the ANNOTATOR to ADMIN and then granting 'rw' permission
+        is left to the ADMIN.
+
+        Must supply a jwt-token to verify user status and extract `user_id`.
+
+        Raises:
+            400 Error if message_type is not specified
+            400 Error if message is not 'r_request' or 'rw_request'
+            401 Error if not logged in
+            404 Error if project does not exist
+
+        Returns:
+            201 if success. Will also return `project_id`.
+        """
+
+    req = request.get_json()
+    check_json(req, ['message_type'])  # check missing keys
+
+    project = Project.query.filter_by(id=project_id).first()
+    if project is None:
+        abort(404, 'Project with id=%s does not exist' % project_id)
+
+    current_user = get_jwt_identity()
+    user = User.query.filter_by(username=current_user).first()
+
+    map_code_to_message_type = {
+        'r_request': MessageTypeEnum.READ_ONLY_PERMISSION_REQUEST,
+        'rw_request': MessageTypeEnum.READ_WRITE_PERMISSION_REQUEST,
```

Comment:
Should we consider changing the request parameter name to 'r' and 'rw' since 'request' is already included in the endpoint?

Reply:
Initially I was worried 'r' and 'rw' would be misleading if in the future we are not only dealing with project request-type of messages. But yeah for now leaving out 'request' might make more sense.